presto-native-execution/presto_cpp/main/types/FunctionMetadata.cpp

```diff
@@ -0,0 +1,210 @@
+
+    updateFunctionMetadata(name, *signature, jsonMetadata);
+    jsonMetadataList.emplace_back(jsonMetadata);
+  }
+  return jsonMetadataList;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata>
+buildScalarJsonMetadata(
+    const std::string& name,
+    const std::vector<const FunctionSignature*>& signatures) {
+  std::vector<protocol::JsonBasedUdfFunctionMetadata> jsonMetadataList;
+  jsonMetadataList.reserve(signatures.size());
+  const protocol::FunctionKind kind = protocol::FunctionKind::SCALAR;
+
+  for (const auto& signature : signatures) {
+    protocol::JsonBasedUdfFunctionMetadata jsonMetadata;
+    jsonMetadata.functionKind = kind;
+    jsonMetadata.routineCharacteristics =
+        getRoutineCharacteristics(*signature, name, kind);
+
+    updateFunctionMetadata(name, *signature, jsonMetadata);
+    jsonMetadataList.emplace_back(jsonMetadata);
+  }
+  return jsonMetadataList;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata>
+buildWindowJsonMetadata(
+    const std::string& name,
+    const std::vector<FunctionSignaturePtr>& signatures) {
+  std::vector<protocol::JsonBasedUdfFunctionMetadata> jsonMetadataList;
+  jsonMetadataList.reserve(signatures.size());
+  const protocol::FunctionKind kind = protocol::FunctionKind::WINDOW;
+
+  for (const auto& signature : signatures) {
+    protocol::JsonBasedUdfFunctionMetadata jsonMetadata;
+    jsonMetadata.functionKind = kind;
+    jsonMetadata.routineCharacteristics =
+        getRoutineCharacteristics(*signature, name, kind);
+
+    updateFunctionMetadata(name, *signature, jsonMetadata);
+    jsonMetadataList.emplace_back(jsonMetadata);
+  }
+  return jsonMetadataList;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata> getFunctionMetadata(
+    const std::string& name) {
+  if (auto aggregateSignatures = getAggregateFunctionSignatures(name)) {
+    return buildAggregateJsonMetadata(name, aggregateSignatures.value());
+  } else if (auto windowSignatures = getWindowFunctionSignatures(name)) {
+    return buildWindowJsonMetadata(name, windowSignatures.value());
+  } else {
+    auto signatures = getFunctionSignatures();
+    if (signatures.find(name) != signatures.end()) {
+      return buildScalarJsonMetadata(name, signatures.at(name));
+    }
+  }
+  VELOX_UNREACHABLE("Function kind for {} cannot be determined", name);
+}
+
+} // namespace
+
+json getJsonMetadataForFunction(const std::string& name) {
+  auto metadataList = getFunctionMetadata(name);
+  auto functionName = getFunctionName(name);
+  json j = json::array();
+  json tj;
+  for (const auto& metadata : metadataList) {
+    protocol::to_json(tj, metadata);
+    j.emplace_back(tj);
+  }
+  return j;
+}
+
+json getJsonFunctionMetadata() {
+  // Get all registered aggregate, window, and scalar function names from
+  // FunctionRegistry.
+  auto functions = getSortedAggregateNames();
+  auto scalarFunctions = getSortedScalarNames();
```

Comment:
Is it possible to use std::vector::insert method instead of writing individual for loops ?

Reply:
Updated, thanks.